pkg/liquidity-source/balancer-v2/stable/pool_simulator.go

```diff
@@ -150,17 +232,316 @@ func (s *PoolSimulator) CalcAmountOut(params pool.CalcAmountOutParams) (*pool.Ca
+) error {
+	if swapFeePercentage.IsZero() {
+		return nil
+	}
+
+	var chosenTokenIndex = 0
+	maxBalance := balances[0]
+	for i := 1; i < len(balances); i++ {
+		if balances[i].Gt(maxBalance) {
+			chosenTokenIndex = i
+			maxBalance = balances[i]
+		}
+	}
+
+	dueProtocolFeeAmount, err := math.StableMath.CalcDueTokenProtocolSwapFeeAmount(amplificationParameter, balances,
+		invariant, chosenTokenIndex, swapFeePercentage)
+	if err != nil {
+		return err
+	}
+
+	balances[chosenTokenIndex].Sub(balances[chosenTokenIndex], dueProtocolFeeAmount)
+
+	return nil
+}
+
+// https://etherscan.io/address/0x06df3b2bbb68adc8b0e302443692037ed9f91b42#code#F6#L46
+func (s *PoolSimulator) CalcAmountOut(params pool.CalcAmountOutParams) (*pool.CalcAmountOutResult, error) {
+	if s.paused {
+		return nil, ErrPoolPaused
+	}
+
+	if s.poolSpec != poolSpecializationGeneral && len(s.Info.Tokens) != 2 {
+		return nil, ErrNotTwoTokens
+	}
+
+	tokenAmountIn, tokenOut := params.TokenAmountIn, params.TokenOut
+	amountIn, overflow := uint256.FromBig(tokenAmountIn.Amount)
+	if overflow {
+		return nil, ErrInvalidAmountIn
+	}
+
+	indexIn, indexOut := s.GetTokenIndex(tokenAmountIn.Token), s.GetTokenIndex(tokenOut)
+
+	if (indexIn < 0 || indexOut < 0) && !s.batchSwapEnabled {
+		return nil, ErrBatchSwapDisabled
+	}
+
+	if indexIn >= 0 && indexOut >= 0 {
+		return s.swapDirect(tokenAmountIn.Token, tokenOut, amountIn)
+	}
+
+	if indexIn < 0 && indexOut >= 0 {
+		return s.swapFromBase2Main(tokenAmountIn.Token, tokenOut, amountIn)
+	}
+
+	if indexIn >= 0 && indexOut < 0 {
+		return s.swapFromMain2Base(tokenAmountIn.Token, tokenOut, amountIn)
+	}
+
+	return s.swapBetweenBasePools(tokenAmountIn.Token, tokenOut, amountIn)
+}
+
+func (s *PoolSimulator) swapDirect(tokenIn, tokenOut string, amountIn *uint256.Int) (*pool.CalcAmountOutResult, error) {
+	amountOut, err := s.OnSwap(tokenIn, tokenOut, amountIn)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.buildSwapResult(tokenOut, amountOut, nil), nil
+}
+
+func (s *PoolSimulator) swapFromBase2Main(tokenIn, tokenOut string, amountIn *uint256.Int) (*pool.CalcAmountOutResult, error) {
+	basePool, err := s.getBasePool(tokenIn)
+	if err != nil {
+		return nil, err
+	}
+
+	bptToken := basePool.GetAddress()
+
+	var (
+		hops = make([]shared.Hop, 0, 2)
+
+		bptAmount *uint256.Int
+		joinIndex *big.Int
+	)
+
+	switch basePool.GetType() {
+	case composablestable.DexType:
+		bptAmount, err = basePool.OnSwap(tokenIn, bptToken, amountIn)
+	default:
+		bptAmount, err = basePool.OnJoin(tokenIn, amountIn)
+		joinIndex = shared.PackJoinExitIndex(shared.PoolJoin, 0)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	hops = append(hops, shared.Hop{
+		PoolId:        basePool.GetPoolId(),
+		Pool:          basePool.GetAddress(),
+		TokenIn:       tokenIn,
+		TokenOut:      bptToken,
+		AmountIn:      amountIn.ToBig(),
+		AmountOut:     bptAmount.ToBig(),
+		JoinExitIndex: joinIndex,
+	})
+
+	amountOut, err := s.OnSwap(tokenOut, bptToken, bptAmount)
```

Comment:
Is the token order flipped?
```suggestion
	amountOut, err := s.OnSwap(bptToken, tokenOut, bptAmount)
```